hack/generator/pkg/codegen/pipeline_create_storage_types.go

```diff
@@ -0,0 +1,234 @@
+	}
+
+	return result
+}
+
+type StorageTypeFactory struct {
+	types               astmodel.Types
+	propertyConversions []propertyConversion
+	visitor             astmodel.TypeVisitor
+}
+
+// A property conversion accepts a property definition and optionally applies a conversion to make
+// the property suitable for use on a storage type. Conversions return nil if they decline to
+// convert, deferring the conversion to another.
+type propertyConversion = func(property *astmodel.PropertyDefinition, ctx StorageTypesVisitorContext) (*astmodel.PropertyDefinition, error)
+
+func (factory *StorageTypeFactory) visitTypeName(_ *astmodel.TypeVisitor, name astmodel.TypeName, ctx interface{}) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+
+	// Resolve the type name to the actual referenced type
+	actualDefinition, actualDefinitionFound := factory.types[name]
+
+	// Check for property specific handling
+	if vc.property != nil && actualDefinitionFound {
+
+		if et, ok := actualDefinition.Type().(*astmodel.EnumType); ok {
+			// Property type refers to an enum, so we use the base type instead
+			return et.BaseType(), nil
+		}
+	}
+
+	// Map the type name into our storage namespace
+	localRef, ok := name.PackageReference.AsLocalPackage()
+	if !ok {
+		return name, nil
+	}
+
+	storageRef := astmodel.MakeStoragePackageReference(localRef)
+	visitedName := astmodel.MakeTypeName(storageRef, name.Name())
+	return visitedName, nil
+}
+
+func (factory *StorageTypeFactory) visitObjectType(
+	_ *astmodel.TypeVisitor,
+	object *astmodel.ObjectType,
+	ctx interface{}) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+	oc := vc.forObject(object)
+
+	var errs []error
+	properties := object.Properties()
+	for i, prop := range properties {
+		p, err := factory.makeStorageProperty(prop, oc)
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			properties[i] = p
+		}
+	}
+
+	if len(errs) > 0 {
+		err := kerrors.NewAggregate(errs)
+		return nil, err
+	}
+
+	ot := astmodel.NewObjectType().WithProperties(properties...)
+	return astmodel.NewStorageType(*ot), nil
+}
+
+// makeStorageProperty applies a conversion to make a variant of the property for use when
+// serializing to storage
+func (factory *StorageTypeFactory) makeStorageProperty(
+	prop *astmodel.PropertyDefinition,
+	pc StorageTypesVisitorContext) (*astmodel.PropertyDefinition, error) {
+	for _, conv := range factory.propertyConversions {
+		p, err := conv(prop, pc.forProperty(prop))
+		if err != nil {
+			// Something went wrong, return the error
+			return nil, err
+		}
+		if p != nil {
+			// We have the conversion we need, return it promptly
+			return p, nil
+		}
+	}
+
+	return nil, fmt.Errorf("failed to find a conversion for property %v", prop.PropertyName())
+}
+
+// Preserve properties required by the KubernetesResource interface unchanged as they're always required
+func (factory *StorageTypeFactory) preserveKubernetesResourceStorageProperties(
```

Comment:
One minor note for later (I've noted it in my notes), it's possible we can make it so that _only_ the storage types implement the `KubernetesResource` interface, since I _think_ that those are the types that the controller is going to operate on?

Reply:
Hmm, but it will have to convert to the actual APIVersion type in question before serializing to/from Azure... I'll think about this some more.